src/main/kotlin/resource/UserResource.kt

```diff
@@ -34,6 +57,76 @@ fun Route.user(userService: UserService) {
                 call.respond("You are missing one or multiple parameters")
             }
         }
+
+        webSocket("/balance") {
+            try {
+                val user = TokenManager.verifyTokenAndRetrieveUser(call.parameters)
+                val balanceManager = userService.buildBalanceManager(user)
+                val balanceIncreaseRateManager = BalanceIncreaseRateManager(user)
+
+                SessionSemaphore.acquireBalanceSession(user)
+
+                try {
+                    while (true) {
+                        balanceIncreaseRateManager.increaseBalanceBasedOnIncreaseRate(balanceManager)
+
+                        val currentBalance = balanceManager.retrieveCurrentBalance()
+                        outgoing.send(Frame.Text(currentBalance.toString()))
+
+                        delay(TimeUnit.SECONDS.toMillis(Scheduler.BalanceIncreaseTimeoutInSeconds))
+                    }
+                } catch (exception: Exception) {
+                    close(CloseReason(CloseReason.Codes.UNEXPECTED_CONDITION, message = WebSocketClosedByExceptionMessage))
+                } finally {
+                    balanceManager.syncCurrentBalanceToDatabase()
+                    SessionSemaphore.releaseBalanceSession(user)
+                }
+            } catch (exception: TokenExpiredException) {
+                close(CloseReason(CloseReason.Codes.CANNOT_ACCEPT, message = exception.message))
+            } catch (exception: TokenMissingException) {
+                close(CloseReason(CloseReason.Codes.CANNOT_ACCEPT, message = exception.message))
+            } catch (exception: SessionLockAlreadyAcquired) {
+                close(CloseReason(CloseReason.Codes.VIOLATED_POLICY, message = exception.message))
+            }
+        }
+
+        webSocket("/clicking") {
+            try {
+                val user = TokenManager.verifyTokenAndRetrieveUser(call.parameters)
+                val balanceManager = userService.buildBalanceManager(user)
+
+                SessionSemaphore.acquireClickingSession(user)
+
+                try {
```

Comment:
Method logic is hard to understand. I would suggest to refactor everything inside this try statement (including the try itself) to a separate method having a descriptive name.